test/Icicle/Test/Core/Exp/Eval.hs

```diff
@@ -44,14 +45,28 @@ prop_prefixlet =
 
 -- Constant evaluates to constant. How quaint.
 -- =====================
+
 prop_const i =
  eval0 evalPrim (constI i :: X.Exp Var) == Right (VBase $ VInt i)
 
 -- And likewise, putting anything that typechecks before the constant still evalutes fine
 -- =====================
+
 prop_constprefix i =
  withTypedExp $ \x _ -> eval0 evalPrim (XLet (fresh 0) x (constI i)) == Right (VBase $ VInt i)
 
+-- Applying small-step dynamics should eventually get to the same result as
+-- applying big-step.
+-- =====================
+
+prop_small_transition =
+ withTypedExp $ \x _ -> iterateEval x == hush (eval0 evalPrim x)
+ where iterateEval x
+         | Right z  <- ES.eval evalPrim x, z /= x = iterateEval z
+         | XValue{} <- x = hush $ eval0 evalPrim x
+         | otherwise = Nothing
+       hush (Left _)  = Nothing
+       hush (Right x) = Just x
```

Comment:
We have our own [rightToMaybe](https://github.com/ambiata/p/blob/265cd5c3d5a91a332aedfd950d80bc1f9e9fc5f2/src/P/Either.hs#L17) if you prefer. :)